### PR TITLE
[Improvement-3690][common] Get the native IP policy problem (获取本机ip策略问题 )

### PR DESCRIPTION
**Pull request by felix-thinkingdata**
## *Tips*
- *Thanks very much for contributing to Apache DolphinScheduler.*
- *Please review https://dolphinscheduler.apache.org/en-us/community/index.html before opening a pull request.*

## What is the purpose of the pull request

Improvement #3690

The current policy is to take the first one, but the first test is not fixed。This is my debug screenshot.

---
当前策略是取第一个，但是第一个测试下来也是不固定的，这个是我调试截图。
image
image

The problem is that on the server, the acquired IP is the external network card IP.External network control is more strict, so it is not a full port through. Therefore, it is recommended to get IP of Intranet card first.So make logical optimizations in your code.

---
遇到的问题是服务器上，获取的ip为外网网卡ip.外网控制比较严格，所以不是全端口打通。所以，建议优先获取内网网卡ip。
所以做出代码中的逻辑优化。

### CODE REVIEW

**Comment by yangyichao-mango** [comment]
Please add test case to make sure this will success run in github docker.

**Review by CalvinKirs — COMMENTED** [review]
In my opinion, preferential selection of internal network IP may not be suitable for some scenarios. In some cases, we need to obtain external network IP. Therefore, it is recommended to set a switch. By default, internal network IP is selected, but users can specify whether to give priority to Obtain the external network IP.

**Comment by felix-thinkingdata** [comment]
> In my opinion, preferential selection of internal network IP may not be suitable for some scenarios. In some cases, we need to obtain external network IP. Therefore, it is recommended to set a switch. By default, internal network IP is selected, but users can specify whether to give priority to Obtain the external network IP.


这种情况可以考虑在内，我加一个快关吧。默认内网优先。如果开关开启就把这个策略关闭，还是加一个外网优先？之前取第一个的策略是不合理的。
This can be taken into account. Let me add a quick switch. Default Intranet priority. Turn this policy off if the switch is on, or add an extranet first? The previous strategy of taking the first one was not reasonable.

**Comment by felix-thinkingdata** [comment]
> > In my opinion, preferential selection of internal network IP may not be suitable for some scenarios. In some cases, we need to obtain external network IP. Therefore, it is recommended to set a switch. By default, internal network IP is selected, but users can specify whether to give priority to Obtain the external network IP.
> 
> 这种情况可以考虑在内，我加一个快关吧。默认内网优先。如果开关开启就把这个策略关闭，还是加一个外网优先？之前取第一个的策略是不合理的。
> This can be taken into account. Let me add a quick switch. Default Intranet priority. Turn this policy off if the switch is on, or add an extranet first? The previous strategy of taking the first one was not reasonable.

考虑到有的云服务器只有一块外网网卡的情况，需要做一个兼容。
或者有的用户，需要优先获取外网网卡，做一个开关逻辑。
Considering that some cloud servers only have one external network card, need to do a compatibility.
Or some users, need to get the priority of the external network card, do a switch logic.

**Comment by felix-thinkingdata** [comment]
> In my opinion, preferential selection of internal network IP may not be suitable for some scenarios. In some cases, we need to obtain external network IP. Therefore, it is recommended to set a switch. By default, internal network IP is selected, but users can specify whether to give priority to Obtain the external network IP.
帮忙起个名字 ，哈哈
The name of this switch in the configuration file, please give it a name

**Comment by CalvinKirs** [comment]
> .

In fact, in the original design, users can display the specified ip

For scenarios where there are real multiple network cards and IP coexistence between internal and external networks, you cannot work hard on the framework side, and users also need to do something. There is no near perfect solution

事实上，原有的设计上优先选取用户所配置的网卡。

对于多网卡，内外网IP共存的场景，不能只是框架在做努力，用户也需要做一些事。没有近乎完美的方案。

**Comment by CalvinKirs** [comment]
> > > In my opinion, preferential selection of internal network IP may not be suitable for some scenarios. In some cases, we need to obtain external network IP. Therefore, it is recommended to set a switch. By default, internal network IP is selected, but users can specify whether to give priority to Obtain the external network IP.
> > 
> > 
> > 这种情况可以考虑在内，我加一个快关吧。默认内网优先。如果开关开启就把这个策略关闭，还是加一个外网优先？之前取第一个的策略是不合理的。
> > This can be taken into account. Let me add a quick switch. Default Intranet priority. Turn this policy off if the switch is on, or add an extranet first? The previous strategy of taking the first one was not reasonable.
> 
> 考虑到有的云服务器只有一块外网网卡的情况，需要做一个兼容。
> 或者有的用户，需要优先获取外网网卡，做一个开关逻辑。
> Considering that some cloud servers only have one external network card, need to do a compatibility.
> Or some users, need to get the priority of the external network card, do a switch logic.

In fact, in the original design, users can display the specified ip

For scenarios where there are real multiple network cards and IP coexistence between internal and external networks, you cannot work hard on the framework side, and users also need to do something. There is no near perfect solution

事实上，原有的设计上优先选取用户所配置的网卡。

对于多网卡，内外网IP共存的场景，不能只是框架在做努力，用户也需要做一些事。没有近乎完美的方案。

**Comment by CalvinKirs** [comment]
Our default plan is to choose the internal network first, and then the external network. The user can display the designated intranet or extranet. If the user shows that the external network is specified, an error will be reported directly if the external network is not available

我们默认方案优先选择内网，其次选择外网。用户可以显示指定选择内网或者外网。如果用户显示指定了外网，拿不到外网的情况下就直接报错 @qiaozhanwei  hi,do you have any good advice

**Comment by yangyichao-mango** [comment]
+1


------------------ 原始邮件 ------------------
发件人: CalvinKirs <notifications@github.com&gt;
发送时间: 2020年9月9日 10:43
收件人: apache/incubator-dolphinscheduler <incubator-dolphinscheduler@noreply.github.com&gt;
抄送: Yichao Yang <1048262223@qq.com&gt;, Comment <comment@noreply.github.com&gt;
主题: 回复：[apache/incubator-dolphinscheduler] [Improvement][COMMON-NetUtils] Get the native IP policy problem (获取本机ip策略问题 ) (#3695)





 
Our default plan is to choose the internal network first, and then the external network. The user can display the designated intranet or extranet. If the user shows that the external network is specified, an error will be reported directly if the external network is not available
 
我们默认方案优先选择内网，其次选择外网。用户可以显示指定选择内网或者外网。如果用户显示指定了外网，拿不到外网的情况下就直接报错 @qiaozhanwei  hi,do you have any good advice

**Comment by felix-thinkingdata** [comment]
@CalvinKirs CalvinKirs  Development has been completed

File: dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/utils/NetUtils.java [hunk 1]
```diff
@@ -227,4 +230,72 @@ private static boolean isSpecifyNetworkInterface(NetworkInterface networkInterfa
...
+    private static NetworkInterface findOuterAddress(List<NetworkInterface> validNetworkInterfaces) {
+        NetworkInterface networkInterface = null;
+        for (NetworkInterface ni : validNetworkInterfaces) {
+            Enumeration<InetAddress> address = ni.getInetAddresses();
+            while (address.hasMoreElements()) {
+                InetAddress ip = address.nextElement();
+                if (!ip.isSiteLocalAddress()
+                        && !ip.isLoopbackAddress()
+                        && !ip.getHostAddress().contains(":")) {
+                    networkInterface = ni;
```
Comment (CalvinKirs) [inline dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/utils/NetUtils.java:294]: Hi, there may be a little problem here, we need to consider the IPV6 situation,

Comment (felix-thinkingdata) [reply dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/utils/NetUtils.java]: yes  i will support ipv6  --x:--x:--x:--x:--x:--x:--x:--x,remove  && !ip.getHostAddress().contains(":")

**Review by CalvinKirs — APPROVED** [review]
looks good to me,please you review @dailidong @qiaozhanwei

File: dolphinscheduler-common/src/main/resources/common.properties [hunk 2]
```diff
@@ -72,3 +72,7 @@ kerberos.expire.time=2
 # datasource encryption salt
 datasource.encryption.enable=false
 datasource.encryption.salt=!@#$%^&*
+
+#Network IP gets priority, default inner outer
```
Comment (yangyichao-mango) [inline dolphinscheduler-common/src/main/resources/common.properties:76]:
```suggestion
# Network IP gets priority, default inner outer
```

Format the description as the others.

**Review by yangyichao-mango — CHANGES_REQUESTED** [review]

**Comment by yangyichao-mango** [comment]
Please change the title to `[Improvement-3690][common] Get the native IP policy problem (获取本机ip策略问题 )`

**Review by yangyichao-mango — APPROVED** [review]
LGTM

**Review by qiaozhanwei — APPROVED** [review]
+1

**Pull request merged**